optproviderupgrade/optproviderupgrade.go

```diff
@@ -38,6 +38,9 @@ func CacheDir(elem ...string) PreviewProviderUpgradeOpt {
 	})
 }
 
+// NewSourcePath sets the path to new source code to use for the target version of the upgrade.
+// If not set, the original pulumitest program source is used.
+// This is useful for where it's expected for a user to perform code changes during a migration.
```

Comment:
"Like the `source` argument to `NewPulumiTest`, the path can be absolute, or relative to the directory the test source is in."